File: my-folder/0494-target-sum/solution.go
func findTargetSumWays(nums []int, target int) int {
    dp := make(map[[2]int]int)

    var dfs func(int, int) int
    dfs = func(i int, sum int) int {
        if i == len(nums) {
            if sum == target {
                return 1
            }
            return 0
        }

        key := [2]int{i, sum}

        if val, ok := dp[key]; ok {
            return val
        }

        dp[key] = dfs(i+1, sum + nums[i]) + dfs(i+1, sum - nums[i])
        return dp[key]
    }

    return dfs(0, 0)
}

File: my-folder/0494-target-sum/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, target: i32) -> i32 {
        let mut dp: HashMap<i32, i32> = HashMap::new();
        dp.insert(0, 1);

        for &num in &nums {
            let mut next_dp: HashMap<i32, i32> = HashMap::new();

            for (&total, &count) in &dp {
                *next_dp.entry(total + num).or_insert(0) += count;
                *next_dp.entry(total - num).or_insert(0) += count;
            }
            dp = next_dp;
        }

        *dp.get(&target).unwrap_or(&0)
    }
}
